sdk/cosmos/azure-cosmos-encryption/src/main/java/com/azure/cosmos/implementation/encryption/AeadAes256CbcHmac256Algorithm.java

```diff
@@ -0,0 +1,382 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.cosmos.implementation.encryption;
+
+import com.azure.cosmos.implementation.encryption.api.CosmosEncryptionAlgorithm;
+import com.azure.cosmos.implementation.encryption.api.DataEncryptionKey;
+import com.azure.cosmos.implementation.encryption.api.EncryptionType;
+
+
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+/**
+ * This class implements authenticated encryption algorithm with associated data as described in
+ * http://tools.ietf.org/html/draft-mcgrew-aead-aes-cbc-hmac-sha2-05 - specifically this implements
+ * AEAD_AES_256_CBC_HMAC_SHA256 algorithm.
+ * This (and AeadAes256CbcHmac256EncryptionKey) implementation for Cosmos DB is same as the existing
+ * SQL client implementation with StyleCop related changes - also, we restrict to randomized encryption to start with.
+ */
+class AeadAes256CbcHmac256Algorithm implements DataEncryptionKey {
+
+    public final static String ALGORITHM_NAME = "AEAD_AES_256_CBC_HMAC_SHA256";
```

Comment:
Is the implementation leveraged from other libraries? 
If those implementations are in same shared repo, any possibility of avoiding the code copy?

Reply:
I didn't find any lib providing this out of the box to be used as a lib. I will look again after the PR merges if we can re-use any other lib.